### ISSUE TITLE
Error: [$rootScope:infdig] http://errors.angularjs.org/1.4.1/$rootScope

### ISSUE THREAD

**Issue by levalencia** [AzureAD/azure-activedirectory-library-for-js#145]
When I enable azure ad authentication with this lines:
  requireADLogin: true,

and htis

   adalAuthenticationServiceProvider.init(
         {
             instance: 'https://login.microsoftonline.com/',
             tenant: 'mysaasapp.onmicrosoft.com',
             clientId: '33e037a7-b1aa-42ab-9693-6c22d01ca338',
             extraQueryParameter: 'nux=1'
             //cacheLocation: 'localStorage', // enable this for IE, as sessionStorage does not work for localhost.
         },
         $httpProvider
         );

I get the error above.

If I remove the ADAL JS lines, then my app works, so I assume the bug its on the ADAL JS side.
I can share my project via email if I needed because I cant share it publicly

Authentication however still works, but its a very annoying error every time I debug on visual studio

http://stackoverflow.com/questions/30993860/error-rootscopeinfdig-http-errors-angularjs-org-1-4-1-rootscope

**Comment by levalencia** [comment]
BTW, Authentication works only in IE 11, on chrome and firefox there is still an infinite loop and browsers crashed

**Comment by sfgadjo** [comment]
I am having a similar issue when there is a stale session. If I leave the browser open overnight, for example, then come back in the morning and do some action (without doing a page refresh) I get an infinite loop which freezes up the browser. I have only tested this on Chrome so far. Any updates on the cause?

**Comment by fhellwig** [comment]
I have observed this behavior with Chrome, but *not* with Firefox. Strange.

**Comment by mmaitre314** [comment]
I hit the same issue. It looks like watchers for URL changes enter an infinite loop. In ended up using adal.js directly to have better control on URL updates. The pound sign returned after redirect was also an issue and using html5 mode did not help (hit another bug where adal angular doesn't work in IE when using html5 mode).

**Comment by smastous** [comment]
I have spent the better part of 3 days messing with this, and here is what I have found so far.  Short story is that this library does not play nice with ui router if you have more than a simple configuration.  

I have a very similar setup to what you put in the stack overflow post with an abstract state as the root and child states underneath it.  If you have $urlRouterProvider.otherwise set up, it will always go into an infinite loop to the point where you have to end task on Chrome.
What I've noticed: 
* If you comment out the $urlRouterProvider.otherwise, it will work fine but obviously will no longer route you to the state you want when you hit the root url.
* If you comment out the adal code, it will work fine even if you have the otherwise.
* If you type the full url to the child state in the browser, it will work fine.  It only seems to break when you directly hit the root url of the page with an otherwise route configured.

Its difficult to debug because it happens as soon as the page loads, but from what I'm seeing it might actually be ui router going into the loop because of how adal is rewriting the url along with timing issues.

This code will exhibit the problem: 
```
    $stateProvider
        .state('home', {
            templateUrl: 'app/main/main.html',
            abstract: true,
            requireADLogin: true,
            controller: 'mainController',
            controllerAs: 'vm'
        })
        .state('home.child', {
            url: '/child',
            templateUrl: 'app/child/child.html',
            requireADLogin: true
        })
        .state('home.child2', {
            url: '/child2',
            templateUrl: 'app/child2/child2.html',
            requireADLogin: true
        });

    //This causes an infinite loop to the point that you have to end task Chrome
    //$urlRouterProvider.otherwise('/child');

    //This shows that the otherwise function is invoked multiple times.  I think ui router routes to the right
    //state before the adal call gets made.  adal's url tampering removes your state's #/ section from the 
    //url, which causes the otherwise block to run again but also go into an infinite loop for some reason.
    $urlRouterProvider.otherwise(function($inject, $location) {
        console.log($location.$$path);
    });
```

**Comment by smastous** [comment]
A bit more research led me to https://github.com/angular-ui/ui-router/issues/600.  This is certainly an interplay with ui router and adal url rewriting.  If I comment out the adal code, it works, but it also looks like ui router has problems with routing scenarios as described above.

I "fixed" the problem by doing the following:

```
    $urlRouterProvider.otherwise( function($injector, $location) {
        var $state = $injector.get("$state");
        $state.go("whatever.state");
    });
```
instead of: 

```
$urlRouterProvider.otherwise('/someStateUrl');
```

You will still get infinite digest errors in the console, but at least it wont hard lock Chrome anymore, and it seems to work as it should with authentication.

**Comment by venkateshappala** [comment]
I am also facing same issue. I have simple HTML with Login and Logout buttons. Attached HTML and JS files. Please help me!!
[ADAL-Angular.zip](https://github.com/AzureAD/azure-activedirectory-library-for-js/files/101120/ADAL-Angular.zip)

**Comment by CodyMorris** [comment]
I'm having this issue too

**Comment by PavelPikat** [comment]
:+1:

**Comment by narayanmyidea** [comment]
It is not complete solution, Please test it before use. I know still it has some issues when you sigin from multiple tabs.
1. adal.js ,add your own method

```
AuthenticationContext.prototype.customIsAuthenticated = function () {
    var isAuthenticated = true;
    var idtoken = this._getItem(this.CONSTANTS.STORAGE.IDTOKEN);
    if (idtoken === undefined || idtoken===null || idtoken === "")
        return false;
```

    return isAuthenticated;
};
2. adal.angular.js ,inside ProtectedResourceInterceptor,add 

 `if (authService.customIsAuthenticated() === false) {
                                if (authService.userInfo.isAuthenticated === true) {
                                    authService.logOut();
                                    window.location.reload(true);
                                    return $q.reject();
                                }
                            }`

**Comment by tushargupta51** [comment]
Have people tried the workaround that @smastous suggested above:

    $urlRouterProvider.otherwise( function($injector, $location) {
        var $state = $injector.get("$state");
        $state.go("whatever.state");  // this is the state that is passed in the otherwise(...) option.
    });

**Comment by sfgadjo** [comment]
@tushargupta51 I have tried it and it didn't make a difference for me.

**Comment by nigel-dewar** [comment]
I tried this 

$urlRouterProvider.otherwise( function($injector, $location) {
    var $state = $injector.get("$state");
    $state.go("whatever.state");  // this is the state that is passed in the otherwise(...) option.
});

but to no avail....

**Comment by tushargupta51** [comment]
@nigel-dewar Can you share a repro?

**Comment by nigel-dewar** [comment]
Hi @tushargupta51 , I can't right now but will do.

Thought I'd just let you know that I am using adal version 1.7 and not 1.8.

Has this infinite loop issue been fixed in the latest version?

**Comment by tushargupta51** [comment]
@nigel-dewar No, its still an open issue.

**Comment by nigel-dewar** [comment]
Oh ok that is a bummer.

**Comment by nigel-dewar** [comment]
Is the recommended work around at this stage Not to use ui-router then? and just use ng-router? ng-router works with no issues right everyone?

**Comment by tushargupta51** [comment]
@nigel-dewar This
````$urlRouterProvider.otherwise( function($injector, $location) {
var $state = $injector.get("$state");
$state.go("whatever.state"); // this is the state that is passed in the otherwise(...) option.
});````

has worked for some people but it does not seem to be working for you. If you can share a simple repro, I can take a look and see if there is a possible workaround.

**Comment by nigel-dewar** [comment]
Thanks mate, I will give this another shot and see what happens.

Cheers

Nigel

Sent from Samsung Mobile

<div>-------- Original message --------</div><div>From: tushar gupta <notifications@github.com> </div><div>Date:02/04/2016  13:08  (GMT+12:00) </div><div>To: AzureAD/azure-activedirectory-library-for-js <azure-activedirectory-library-for-js@noreply.github.com> </div><div>Cc: nigel-dewar <nigeldewar@live.com> </div><div>Subject: Re: [AzureAD/azure-activedirectory-library-for-js] Error: [$rootScope:infdig] http://errors.angularjs.org/1.4.1/$rootScope (#145) </div><div>
</div>
@nigel-dewar This
````$urlRouterProvider.otherwise( function($injector, $location) {
var $state = $injector.get("$state");
$state.go("whatever.state"); // this is the state that is passed in the otherwise(...) option.
});````

has worked for some people but it does not seem to be working for you. If you can share a simple repro, I can take a look and see if there is a possible workaround.

---
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub:
https://github.com/AzureAD/azure-activedirectory-library-for-js/issues/145#issuecomment-204613787

**Comment by nigel-dewar** [comment]
Hi Tried this and it does not work.

I guess the code above assumes you are using an abstract state like 'app' and then you are wanting to route directly to a state like app.home for instance
//using parent child state
$urlRouterProvider.otherwise( function($injector, $location) { var $state = $injector.get("$state"); $state.go("app.home"); 

//using state
$urlRouterProvider.otherwise( function($injector, $location) { var $state = $injector.get("$state"); $state.go("home"); 

I have tried both using a parent and child state and using just states alone and it does not work. And besides that I dont want to have an abstract state like app, and route to a child state like app.home.

regardless this fix does not work for me.

I am using angular 1.4.9 with ui-router 0.2.18

referencing this issue here https://github.com/angular-ui/ui-router/issues/600  I can see that the issue is still not fixed.

My plans now are to to stop using ui-router and fall back to using ng-router in hopes I can get some joy.

**Comment by PavelPikat** [comment]
Hi @nigel-dewar 
Can you try something like

` $urlRouterProvider.otherwise(
            function ($injector, $stateParams) {

                var isAdalFrame = window !== window.parent;

                var $state = $injector.get('$state'),
                    $$path = $stateParams.$$path,
                    $$url = $stateParams.$$url,
                    $log = $injector.get('$log');
                
                if ($$path) {
                    if (($$path.indexOf('id_token') !== -1 &&
                        $$url.indexOf('id_token') !== -1) ||
                        ($$path.indexOf('access_token') !== -1 &&
                        $$url.indexOf('access_token') !== -1) ||
                        $$path.indexOf('/error=') !== -1) {
                        // Do nothing if we have adal's token in the url
                    } else if (!isAdalFrame) {
                        $state.go(%YOUR_HOME_ROUTE%);
                    }
                }
                else if (!isAdalFrame) {
                    $state.go(%YOUR_HOME_ROUTE%);
                }
            }
        );`

We got rid of digest looping in our app using this approach.
We also added checks for "access_token" and "id_token" in $locationChangeStart event to preventDefault

**Comment by nigel-dewar** [comment]
Hi @PavelPikat ,

Sorry this does not work either, still getting:

Error: [$rootScope:infdig] http://errors.angularjs.org/1.4.9/$rootScope/infdig?p0=10&p1=%5B%5D
    at Error (native)
    at http://localhost:8000/lib/angular/angular.min.js:6:416
    at r.$get.r.$digest (http://localhost:8000/lib/angular/angular.min.js:132:132)
    at r.$get.r.$apply (http://localhost:8000/lib/angular/angular.min.js:134:236)
    at http://localhost:8000/lib/angular/angular.min.js:20:59
    at Object.e [as invoke] (http://localhost:8000/lib/angular/angular.min.js:39:431)
    at c (http://localhost:8000/lib/angular/angular.min.js:19:482)
    at zc (http://localhost:8000/lib/angular/angular.min.js:20:274)
    at ce (http://localhost:8000/lib/angular/angular.min.js:19:83)
    at HTMLDocument.<anonymous> (http://localhost:8000/lib/angular/angular.min.js:297:355)(anonymous function) @ angular.js:12722$get @ angular.js:9490$get.r.$apply @ angular.js:16361(anonymous function) @ angular.js:1680e @ angular.js:4535c @ angular.js:1678zc @ angular.js:1698ce @ angular.js:1592(anonymous function) @ angular.js:29652n.Callbacks.i @ jquery.min.js:2n.Callbacks.j.fireWith @ jquery.min.js:2n.extend.ready @ jquery.min.js:2J @ jquery.min.js:2
angular.js:38 Uncaught Error: [$rootScope:infdig] http://errors.angularjs.org/1.4.9/$rootScope/infdig?p0=10&p1=%5B%5D

Thank you for trying though.

**Comment by nigel-dewar** [comment]
It seems to happen immediately after a login. Refreshing the app in chrome ( cache turned off ), does not cause any issue.

when logging out, log back in, the error message shown copy/pasted in previous post appears. Just can't get round it.

Here is my app.js if it helps...

/// <reference path="app/modules/mail/mail.html" />
(function () {
    'use strict';

    var app = angular.module('app', [
        'ngSanitize',
        'AdalAngular',
        'ui.bootstrap',
        'ui.calendar',
        'ui.select',
        'ui.router',
        'ngAnimate',
        'ngCkeditor'
    ])
    .constant("constantService", {
        //set to false when go live
        testMode: false,

        frontEndUrl: "http://localhost:8000/",
        // frontEndUrl: "http://myapp.azurewebsites.net/",

        appUrl: "https://localhost:44301/"
        //  appUrl: "https://myapiapp.azurewebsites.net/"
    })
    .config(config);
    //.run(['$route', 'uiSelect2Config', function ($route, uiSelect2Config) {
        //.run([, function () {
        ////$route.reload();
        //    //$browser.baseHref = function () { return "/" };
        //}]);

    function config($stateProvider, $urlRouterProvider, adalAuthenticationServiceProvider, $httpProvider, $locationProvider) {


        $stateProvider
            // Pages Routes ---------------------------------------------------------------------//

            .state("home", {
                url: "/home",
                controller: "homeController",
                controllerAs: "vm",
                templateUrl: "app/modules/home/home.html",
                requireADLogin: true
            })

             .state("news", {
                 url: "/news",
                 controller: "newsController",
                 controllerAs: "vm",
                 templateUrl: "app/modules/news/news.html",
                 requireADLogin: true
             })

             .state("contacts", {
                 url: "/contacts",
                 controller: "contactsController",
                 controllerAs: "vm",
                 templateUrl: "app/modules/contacts/contacts.html",
                 requireADLogin: true
             })

             .state("mentoring", {
                 url: "/mentoring",
                 controller: "mentoringController",
                 controllerAs: "vm",
                 templateUrl: "app/modules/mentoring/mentoring.html",
                 requireADLogin: true
             })

             // App Routes ---------------------------------------------------------------------//
            .state("presenters", {
                url: "/presenters",
                controller: "presentersCtrl",
                controllerAs: "vm",
                templateUrl: "app/modules/presenters/presenters.html",
                requireADLogin: true
            })

            .state("presenter", {
                url: "/presenters/:presenterId",
                controller: "presentersDetailsCtrl",
                controllerAs: "vm",
                templateUrl: "app/modules/presenters/presentersDetails.html",
                requireADLogin: true
            })

             // Get Roles
            .state("roles", {
                url: "/roles",
                controller: "rolesController",
                controllerAs: "vm",
                templateUrl: "app/modules/roles/roles.html",
                requireADLogin: true
            })

        // Get Workshops
            .state("workshops", {
                url: "/workshops/:workshopId",
                controller: "workshopDetailsController",
                controllerAs: "vm",
                templateUrl: "app/modules/workshops/workshopDetails.html",
                requireADLogin: true
            })

            .state("sessionGroup", {
                url: "/workshops/:workshopId/sessionGroups/:sessionGroupId",
                controller: "workshopDetailsController",
                controllerAs: "vm",
                templateUrl: "app/modules/workshops/workshopDetails.html",
                requireADLogin: true
            })

            .state("calendar", {
                url: "/calendar",
                controller: "calendarController",
                controllerAs: "vm",
                templateUrl: "app/modules/calendar/calendar.html",
                requireADLogin: true
            })

            .state("capabilities", {
                url: "/capabilities",
                controller: "capabilitiesController",
                controllerAs: "vm",
                templateUrl: "app/modules/capabilities/capabilities.html",
                requireADLogin: true
            })

             .state("capabilitiesById", {
                 url: "/roles/:roleId/capabilities/:capabilityId",
                 controller: "capabilitiesDetailsController",
                 controllerAs: "vm",
                 templateUrl: "app/modules/capabilities/capabilitiesDetails.html",
                 requireADLogin: true
             })

            .state("capabilitiesByRole", {
                url: "/roles/:roleId/capabilities",
                controller: "capabilitiesController",
                controllerAs: "vm",
                templateUrl: "app/modules/capabilities/capabilities.html",
                requireADLogin: true
            })

             .state("sessionsByWorkshop", {
                 url: "/workshops/:workshopId/sessiongroups/:sessiongroupId",
                 controller: "sessionDetailsController",
                 controllerAs: "vm",
                 templateUrl: "app/modules/sessions/sessionDetails.html",
                 requireADLogin: true
             })

              // Presenters Routes ---------------------------------------------------------------------//

             .state("presentersDashboard", {
                 url: "/presentersdashboard",
                 controller: "presentersDashboard",
                 controllerAs: "vm",
                 templateUrl: "app/modules/presentersdashboard/presentersDashboard.html",
                 requireADLogin: true
             })

            // Admin Routes ---------------------------------------------------------------------//
            
            .state("admin", {
                url: "/admin",
                abstract: true,
                controller: "adminShellController",
                controllerAs: "vm",
                templateUrl: "app/modules/admin/adminshell.html",
                requireADLogin: true
            })

            .state("admin.roles", {
                url: "/roles",
                views: {
                    'tabContent': {
                        controller: "roleAdmin",
                        controllerAs: "vm",
                        templateUrl: "app/modules/admin/roleAdmin.html",
                        requireADLogin: true
                    }
                }
            })

        .state("admin.capabilities", {
            url: "/capabilities",
            views: {
                'tabContent': {
                    controller: "capabilityAdmin",
                    controllerAs: "vm",
                    templateUrl: "app/modules/admin/capabilityAdmin.html",
                    requireADLogin: true
                }
            }
        })

        .state("admin.workshops", {
            url: "/workshops",
            views: {
                'tabContent': {
                    controller: "workshopAdmin",
                    controllerAs: "vm",
                    templateUrl: "app/modules/admin/workshopAdmin.html",
                    requireADLogin: true
                }
            }
        })

        .state("admin.presenters", {
            url: "/presenters",
            views: {
                'tabContent': {
                    controller: "presenterAdmin",
                    controllerAs: "vm",
                    templateUrl: "app/modules/admin/presenterAdmin.html",
                    requireADLogin: true
                }
            }
        });

        //$urlRouterProvider.when('', '/home');

        //$urlRouterProvider.otherwise('/home');

        $urlRouterProvider.otherwise(
               function ($injector, $stateParams) {

                   var isAdalFrame = window !== window.parent;

                   var $state = $injector.get('$state'),
                       $$path = $stateParams.$$path,
                       $$url = $stateParams.$$url,
                       $log = $injector.get('$log');

                   if ($$path) {
                       if (($$path.indexOf('id_token') !== -1 &&
                           $$url.indexOf('id_token') !== -1) ||
                           ($$path.indexOf('access_token') !== -1 &&
                           $$url.indexOf('access_token') !== -1) ||
                           $$path.indexOf('/error=') !== -1) {
                           // Do nothing if we have adal's token in the url
                       } else if (!isAdalFrame) {
                           $state.go('home');
                       }
                   }
                   else if (!isAdalFrame) {
                       $state.go('home');
                   }
               }
       );



        // The endpoints here are resources for cross origin requests.
        var endpoints = {
            "https://graph.microsoft.com": "https://graph.microsoft.com",
            "https://mytenant.sharepoint.com": "https://mytenant.sharepoint.com",
            "https://myapp.azurewebsites.net": "https://mytenant.onmicrosoft.com/SEDAPI",
            "https://localhost:44301": "https://mytenant.onmicrosoft.com/SEDAPI"
        };


        adalAuthenticationServiceProvider.init(
			{
			    tenant: 'MY_TENANT',
			    clientId: 'MY_CLIENT_ID',
			    //extraQueryParameter: 'nux=1',
			    endpoints: endpoints,
			    cacheLocation: 'localStorage'
			},
			$httpProvider
			);
    };
})();

**Comment by nigel-dewar** [comment]
Guys I found the solution for anyone banging head on wall with this, but you wont like it.

ng-router,. no more looping digests.

shame though as ui-router is really nice.

**Comment by PavelPikat** [comment]
@nigel-dewar 
Our app works fine with ui-router. What about locationChangeStart? Have you tried it?

```
function ($rootScope, $location) {
            var isAdalFrame = window !== window.parent;
            $rootScope.$on('$locationChangeStart', function (e) {

                if ($location.path().indexOf('access_token') > -1 ||
                    $location.path().indexOf('id_token') > -1 ||
                    $location.path().indexOf('/error=') === 0 ||
                    isAdalFrame
                )
                {
                    e.preventDefault();
                }
        }
}
```

**Comment by nigel-dewar** [comment]
I haven't tried this sorry but I will give this a shot.

thanks for posting this through.

**Comment by nigel-dewar** [comment]
Mate, your a legend!!! this really solves this issue!!!!

**Comment by tkostynyuk** [comment]
I'm trying to add this code to the config.js and get the error that $state is not defined. Services can't be injected into config functions. Providers can't be injected to run functions. Can you please tell me what is the right place for this code? I see that nigel-dewar have it in the config function.

 $urlRouterProvider.otherwise(function ($inject, $stateParams) {
        var $state = $injector.get("$state");
        $state.go("dashboard.all");
    });

**Comment by tkostynyuk** [comment]
moved down the  $urlRouterProvider.otherwise  and it worked...

**Comment by nigel-dewar** [comment]
Hi tkostynyuk,  although pavelpiket advice worked for me , my app was still be having funny and occasionally freezing. I ended up moving back to ngrouter and have no issues.

**Comment by tkostynyuk** [comment]
Thank you for the reply. Yesterday it worked because i logged into Azure portal in a different tab. Today it's looping again. Maybe upgrading angular and ui.router can help? We have an existing app that uses ui-router and moving to ngrouter is not an option...

**Comment by tkostynyuk** [comment]
Hi Hi @PavelPikat ,
Can you post the code that you added to config and run blocks to fix the issue?

Thanks!

**Comment by PavelPikat** [comment]
@tkostynyuk Just scroll up, it was added in this thread earlier ;)

**Comment by tkostynyuk** [comment]
Is this correct?

in config:

 $urlRouterProvider.otherwise( function($injector, $location) {
        var $state = $injector.get("$state");
        $state.go("whatever.state");
    });

and run:

app.run(
function ($rootScope, $location) {
            var isAdalFrame = window !== window.parent;
            $rootScope.$on('$locationChangeStart', function (e) {

                if ($location.path().indexOf('access_token') > -1 ||
                    $location.path().indexOf('id_token') > -1 ||
                    $location.path().indexOf('/error=') === 0 ||
                    isAdalFrame
                )
                {
                    e.preventDefault();
                }
        }
});

**Comment by tkostynyuk** [comment]
I think i tried all variations that were posted ...

**Comment by PavelPikat** [comment]
@tkostynyuk https://github.com/AzureAD/azure-activedirectory-library-for-js/issues/145#issuecomment-205166820

**Comment by tkostynyuk** [comment]
Thank you. If I leave the browser window  open for a couple hours and then refresh it goes into the loop with this code or without. If i restart it works most of the time - i'm still trying to understand the pattern when it works and when it doesn't.

**Comment by PavelPikat** [comment]
@tkostynyuk Adal's token expiration is 1 hour, I think. Maybe it fails to refresh the token silently after it expires?

**Comment by tkostynyuk** [comment]
Refreshed the page today and reports loaded but the console shows a lot of the same (10 $digest() iterations reached) and some new errors. It works if you don't look at  the console :)

angular.js:80Uncaught Error: [$rootScope:infdig] 10 $digest() iterations reached. Aborting!
Watchers fired in the last 5 iterations: []

PUT https://wabi-us-north-central-redirect.analysis.windows.net/powerbi/metadata/refreshusermetadata 403 (Forbidden)
VM29870 ai.0.js:1 

PUT https://wabi-us-north-central-redirect.analysis.windows.net/powerbi/metadata/refreshusermetadata 403 (Forbidden)

**Comment by tushargupta51** [comment]
We have released 1.0.11 with fixes to some of the digest iterations issue. Please use 1.0.11 and have a look here if you are still running into this issue: https://github.com/AzureAD/azure-activedirectory-library-for-js/wiki/FAQs#q2-my-app-gets-into-an-infinite-loop-sometimes-leading-to-digest-iterations-error. Let us know if you still run into any issues.
